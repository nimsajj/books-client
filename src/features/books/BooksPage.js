import React, { useEffect } from "react";
import { useSelector, useDispatch } from "react-redux";
import { Link } from "react-router-dom";
import {
  GridList,
  GridListTile,
  GridListTileBar,
  IconButton,
  CircularProgress,
  Typography,
} from "@material-ui/core";
import { Info as InfoIcon, Error as ErrorIcon } from "@material-ui/icons";
import { booksStyles as useStyles } from "./styles";
import {
  fetchBooksRequest,
  selectBooks,
  selectBooksError,
  selectBooksStatus,
} from "./redux";

const BooksList = ({ books }) => {
  const classes = useStyles();

  const renderBooks = () =>
    books.map((book) => (
      <GridListTile key={`${book.isbn}`}>
        <img src={book.img} alt={book.title} />
        <GridListTileBar
          title={book.title}
          subtitle={<span>by: unknow author</span>}
          actionIcon={
            <Link to={`/books/${book.id}`}>
              <IconButton
                aria-label={`info about ${book.isbn}`}
                className={classes.icon}
              >
                <InfoIcon />
              </IconButton>
            </Link>
          }
        />
      </GridListTile>
    ));

  return (
    <div className={classes.root}>
      <GridList cellHeight={180} cols={4} className={classes.gridList}>
        {renderBooks()}
      </GridList>
    </div>
  );
};

const useSelectDataToStore = () => {
  const books = useSelector(selectBooks);
  const booksStatus = useSelector(selectBooksStatus);
  const booksError = useSelector(selectBooksError);

  return [books, booksStatus, booksError];
};

const BooksPage = () => {
  const classes = useStyles();
  const dispatch = useDispatch();

  const [books, booksStatus, booksError] = useSelectDataToStore();

  useEffect(() => {
    if (booksStatus === "initial") {
      dispatch(fetchBooksRequest());
    }
  }, [dispatch, booksStatus]);

  const renderCircularProgress = () => {
    if (["initial", "loading"].includes(booksStatus)) {
      return (
        <div className={classes.progress}>
          <CircularProgress size={100} />
        </div>
      );
    }
  };

  const renderError = () => {
    if (booksStatus === "error") {
      return (
        <Typography align="center" color="error" className={classes.error}>
          <ErrorIcon />
          {booksError}
        </Typography>
      );
    }
  };

  return (
    <>
      {renderCircularProgress()}
      {renderError()}
      <BooksList books={books} />
    </>
  );
};

export default BooksPage;
