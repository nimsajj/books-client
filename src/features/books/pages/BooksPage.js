import React, { useEffect } from "react";
import { useSelector, useDispatch } from "react-redux";
import { GridList } from "@material-ui/core";
import {
  fetchBooksRequest,
  selectBooks,
  selectBooksError,
  selectBooksStatus,
} from "../redux";
import { Loader, Warning } from "../../../components/ui";
import { BookListItem } from "../components";
import styles from "./styles.module.css";

const BooksPage = () => {
  const dispatch = useDispatch();

  const books = useSelector(selectBooks);
  const booksStatus = useSelector(selectBooksStatus);
  const booksError = useSelector(selectBooksError);

  useEffect(() => {
    if (booksStatus === "initial") {
      dispatch(fetchBooksRequest());
    }
  }, [dispatch, booksStatus]);

  const renderBooks = () => (
    <div className={styles.root}>
      <GridList cellHeight={180} cols={4} className={styles.gridList}>
        {books.map((book) => (
          <BookListItem key={book.isbn} {...book} />
        ))}
      </GridList>
    </div>
  );

  return (
    <>
      <Loader status={booksStatus} />
      <Warning message={booksError} color="error" />
      {renderBooks()}
    </>
  );
};

export default BooksPage;
