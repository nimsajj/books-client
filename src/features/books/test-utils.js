import React from "react";
import { Provider } from "react-redux";
import createSagaMiddleware from "redux-saga";
import configureMockStore from "redux-mock-store";
import { BrowserRouter as Router } from "react-router-dom";

import renderer from "react-test-renderer";

const mockStore = configureMockStore([createSagaMiddleware()]);

const render = (children, initialState) => {
  const store = mockStore(initialState);

  return renderer.create(
    <Provider store={store}>
      <Router>{children}</Router>
    </Provider>
  );
};

export { render };
