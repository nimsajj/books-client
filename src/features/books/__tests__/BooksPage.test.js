import React from "react";
import { render } from "../test-utils";
import BooksPage from "../BooksPage";

describe("BooksPage component", () => {
  it("should show progress status", () => {
    const initialState = {
      books: {
        status: "initial",
        error: null,
        data: [],
      },
    };

    const component = render(<BooksPage />, initialState);

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should error message", () => {
    const initialState = {
      books: {
        status: "error",
        error: "Error 404",
        data: [],
      },
    };

    const component = render(<BooksPage />, initialState);

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should a book list", () => {
    const initialState = {
      books: {
        status: "succeded",
        error: null,
        data: [
          {
            id: 1,
            title: "Aperiam qui saepe cumque.",
            totalPages: 111,
            rating: "2.48",
            isbn: "8920461264006",
            publishedAt: "2017-01-11T20:08:54+00:00",
            img:
              "https://via.placeholder.com/640x480.png/00ff88?text=books+sed",
          },
          {
            id: 2,
            title: "Soluta repudiandae modi fuga neque enim.",
            totalPages: 819,
            rating: "4.22",
            isbn: "8872543528578",
            publishedAt: "1982-03-25T23:12:37+00:00",
            img:
              "https://via.placeholder.com/640x480.png/003399?text=books+eligendi",
          },
          {
            id: 3,
            title:
              "Quidem praesentium et nobis dignissimos delectus quas quia facere.",
            totalPages: 287,
            rating: "1.69",
            isbn: "2738955534063",
            publishedAt: "1993-02-01T12:25:31+00:00",
            img:
              "https://via.placeholder.com/640x480.png/009988?text=books+maxime",
          },
        ],
      },
    };

    const component = render(<BooksPage />, initialState);

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
