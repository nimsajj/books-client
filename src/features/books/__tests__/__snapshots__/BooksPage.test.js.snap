// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BooksPage component should a book list 1`] = `
<div
  className="makeStyles-root-1"
>
  <ul
    className="MuiGridList-root"
    style={
      Object {
        "margin": -2,
      }
    }
  >
    <li
      className="MuiGridListTile-root"
      style={
        Object {
          "height": 184,
          "padding": 2,
          "width": "25%",
        }
      }
    >
      <div
        className="MuiGridListTile-tile"
      >
        <img
          alt="Aperiam qui saepe cumque."
          src="https://via.placeholder.com/640x480.png/00ff88?text=books+sed"
        />
        <div
          className="MuiGridListTileBar-root MuiGridListTileBar-titlePositionBottom MuiGridListTileBar-rootSubtitle"
        >
          <div
            className="MuiGridListTileBar-titleWrap MuiGridListTileBar-titleWrapActionPosRight"
          >
            <div
              className="MuiGridListTileBar-title"
            >
              Aperiam qui saepe cumque.
            </div>
            <div
              className="MuiGridListTileBar-subtitle"
            >
              <span>
                by: unknow author
              </span>
            </div>
          </div>
          <div
            className="MuiGridListTileBar-actionIcon"
          >
            <a
              href="/books/1"
              onClick={[Function]}
            >
              <button
                aria-label="info about 8920461264006"
                className="MuiButtonBase-root MuiIconButton-root makeStyles-icon-2"
                disabled={false}
                onBlur={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={0}
                type="button"
              >
                <span
                  className="MuiIconButton-label"
                >
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
                    />
                  </svg>
                </span>
              </button>
            </a>
          </div>
        </div>
      </div>
    </li>
    <li
      className="MuiGridListTile-root"
      style={
        Object {
          "height": 184,
          "padding": 2,
          "width": "25%",
        }
      }
    >
      <div
        className="MuiGridListTile-tile"
      >
        <img
          alt="Soluta repudiandae modi fuga neque enim."
          src="https://via.placeholder.com/640x480.png/003399?text=books+eligendi"
        />
        <div
          className="MuiGridListTileBar-root MuiGridListTileBar-titlePositionBottom MuiGridListTileBar-rootSubtitle"
        >
          <div
            className="MuiGridListTileBar-titleWrap MuiGridListTileBar-titleWrapActionPosRight"
          >
            <div
              className="MuiGridListTileBar-title"
            >
              Soluta repudiandae modi fuga neque enim.
            </div>
            <div
              className="MuiGridListTileBar-subtitle"
            >
              <span>
                by: unknow author
              </span>
            </div>
          </div>
          <div
            className="MuiGridListTileBar-actionIcon"
          >
            <a
              href="/books/2"
              onClick={[Function]}
            >
              <button
                aria-label="info about 8872543528578"
                className="MuiButtonBase-root MuiIconButton-root makeStyles-icon-2"
                disabled={false}
                onBlur={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={0}
                type="button"
              >
                <span
                  className="MuiIconButton-label"
                >
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
                    />
                  </svg>
                </span>
              </button>
            </a>
          </div>
        </div>
      </div>
    </li>
    <li
      className="MuiGridListTile-root"
      style={
        Object {
          "height": 184,
          "padding": 2,
          "width": "25%",
        }
      }
    >
      <div
        className="MuiGridListTile-tile"
      >
        <img
          alt="Quidem praesentium et nobis dignissimos delectus quas quia facere."
          src="https://via.placeholder.com/640x480.png/009988?text=books+maxime"
        />
        <div
          className="MuiGridListTileBar-root MuiGridListTileBar-titlePositionBottom MuiGridListTileBar-rootSubtitle"
        >
          <div
            className="MuiGridListTileBar-titleWrap MuiGridListTileBar-titleWrapActionPosRight"
          >
            <div
              className="MuiGridListTileBar-title"
            >
              Quidem praesentium et nobis dignissimos delectus quas quia facere.
            </div>
            <div
              className="MuiGridListTileBar-subtitle"
            >
              <span>
                by: unknow author
              </span>
            </div>
          </div>
          <div
            className="MuiGridListTileBar-actionIcon"
          >
            <a
              href="/books/3"
              onClick={[Function]}
            >
              <button
                aria-label="info about 2738955534063"
                className="MuiButtonBase-root MuiIconButton-root makeStyles-icon-2"
                disabled={false}
                onBlur={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={0}
                type="button"
              >
                <span
                  className="MuiIconButton-label"
                >
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
                    />
                  </svg>
                </span>
              </button>
            </a>
          </div>
        </div>
      </div>
    </li>
  </ul>
</div>
`;

exports[`BooksPage component should error message 1`] = `
Array [
  <p
    className="MuiTypography-root makeStyles-error-4 MuiTypography-body1 MuiTypography-colorError MuiTypography-alignCenter"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-2h2v2zm0-4h-2V7h2v6z"
      />
    </svg>
    Error 404
  </p>,
  <div
    className="makeStyles-root-1"
  >
    <ul
      className="MuiGridList-root"
      style={
        Object {
          "margin": -2,
        }
      }
    />
  </div>,
]
`;

exports[`BooksPage component should show progress status 1`] = `
Array [
  <div
    className="makeStyles-progress-3"
  >
    <div
      className="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
      role="progressbar"
      style={
        Object {
          "height": 100,
          "width": 100,
        }
      }
    >
      <svg
        className="MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <circle
          className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
          cx={44}
          cy={44}
          fill="none"
          r={20.2}
          strokeWidth={3.6}
          style={Object {}}
        />
      </svg>
    </div>
  </div>,
  <div
    className="makeStyles-root-1"
  >
    <ul
      className="MuiGridList-root"
      style={
        Object {
          "margin": -2,
        }
      }
    />
  </div>,
]
`;
